pkg/scheduler/schedulerdiscovery/discovery.go

```diff
@@ -0,0 +1,132 @@
+	// is no longer available.
+	AddressRemoved(address string)
+}
+
+func NewServiceDiscovery(cfg Config, schedulerAddress string, lookupPeriod time.Duration, component string, receiver Notifications, logger log.Logger, reg prometheus.Registerer) (services.Service, error) {
+	// Since this is a client for the query-schedulers ring, we append "query-scheduler-client" to the component to clearly differentiate it.
+	component = component + "-query-scheduler-client"
+
+	switch cfg.Mode {
+	case ModeRing:
+		return NewRingServiceDiscovery(cfg, component, receiver, logger, reg)
+	default:
+		return NewDNSServiceDiscovery(schedulerAddress, lookupPeriod, receiver)
+	}
+}
+
+func NewDNSServiceDiscovery(schedulerAddress string, lookupPeriod time.Duration, receiver Notifications) (services.Service, error) {
+	return util.NewDNSWatcher(schedulerAddress, lookupPeriod, receiver)
+}
+
+var (
+	activeSchedulersOp = ring.NewOp([]ring.InstanceState{ring.ACTIVE}, nil)
+)
+
+type ringServiceDiscovery struct {
+	services.Service
+
+	ring               *ring.Ring
+	subservicesWatcher *services.FailureWatcher
+	checkPeriod        time.Duration
+	receiver           Notifications
+
+	// Keep track of the addresses that have been discovered and notified so far.
+	notifiedAddresses map[string]struct{}
+}
+
+func NewRingServiceDiscovery(cfg Config, component string, receiver Notifications, logger log.Logger, reg prometheus.Registerer) (services.Service, error) {
+	client, err := NewRingClient(cfg.SchedulerRing, component, logger, reg)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &ringServiceDiscovery{
+		ring:               client,
+		subservicesWatcher: services.NewFailureWatcher(),
+		checkPeriod:        cfg.SchedulerRing.RingCheckPeriod,
+		notifiedAddresses:  make(map[string]struct{}),
+		receiver:           receiver,
+	}
+
+	r.subservicesWatcher.WatchService(r.ring)
+	r.Service = services.NewBasicService(r.starting, r.running, r.stopping)
+
+	return r, nil
+}
+
+func (r *ringServiceDiscovery) starting(ctx context.Context) error {
+	return errors.Wrap(services.StartAndAwaitRunning(ctx, r.ring), "failed to start query-schedulers ring client")
+}
+
+func (r *ringServiceDiscovery) stopping(_ error) error {
+	return errors.Wrap(services.StopAndAwaitTerminated(context.Background(), r.ring), "failed to stop query-schedulers ring client")
+}
+
+func (r *ringServiceDiscovery) running(ctx context.Context) error {
+	ringTicker := time.NewTicker(r.checkPeriod)
+	defer ringTicker.Stop()
+
+	// Notifies the initial state.
+	ringState, _ := r.ring.GetAllHealthy(activeSchedulersOp) // nolint:errcheck
+	r.notifyChanges(ringState)
+
+	for {
+		select {
+		case <-ringTicker.C:
+			ringState, _ := r.ring.GetAllHealthy(activeSchedulersOp) // nolint:errcheck
+			r.notifyChanges(ringState)
+		case <-ctx.Done():
+			return nil
+		case err := <-r.subservicesWatcher.Chan():
+			return errors.Wrap(err, "a subservice of query-schedulers ring-based service discovery has failed")
```

Comment:
When this happens, there is "nobody" watching for this error. (Querier worker nor query-frontend doesn't care about discovery-service failures)

Reply:
> Querier worker nor query-frontend doesn't care about discovery-service failures

You're right, but I think they should. If the service discovery stops working, they stop correctly working too.

Reply:
> but I think they should. If the service discovery stops working, they stop correctly working too.

Agree.

Reply:
Done in https://github.com/grafana/mimir/pull/2957/commits/635c14b0f5e10263c86924eb6dbbc4ab72fe3836.

Reply:
Thanks!